### PR TITLE
Notifier and Listener

### PR DESCRIPTION
**Pull request by craig-cogdill**
A wrapper around Shotgun-Alien queue for notifying multiple processes. The Notifier will fire off a message and wait for a confirmation from all Listeners. The confirmation queue is a Rifle-Vampire queue.

### CODE REVIEW

File: src/Listener.cpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+      shooter.reset();
+   }
+   return shooter;
+}
+
+/*
+ * Reset the member Alien to nullptr
+ */
+void Listener::Reset() {
+   mQueueReader.reset(nullptr);
```
Comment (shwetabhandare) [inline src/Listener.cpp]: What happens to the  mHandshakeQueue = std::move(CreateFeedbackShooter()); that was created in Initialize() ? Is that cleaned up?

File: src/Listener.cpp [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+ * The Alien attempts to read a message from
+ * the queue. It checks to see if it received
+ * anything and returns true if it got the
+ * notification message
+ *
+ * @return true if notification message was received
+ */
+bool Listener::NotificationReceived() {
+   std::vector<std::string> dataFromQueue;
+   mQueueReader->GetShot(getShotTimeout, dataFromQueue);
```
Comment (shwetabhandare) [inline src/Listener.cpp]: Does GetShot always return valid data? What happens if GetShot does not receive data for getShotTimeout?

Comment (KjellKod) [inline src/Listener.cpp]: MessageHasPayload should return false;

@craig-cogdill  is this in a unit test?

File: src/Listener.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+#include <string>
+#include <memory>
+
+class Alien;
+class Rifle;
+
+class Listener {
+   public:
+      Listener() = delete;
+      Listener(const std::string& notificationQueue, const std::string& handshakeQueue, const std::string& program);
```
Comment (KjellKod) [inline src/Listener.h]: I do not like the double construction phase. 

1) Constructor + 2) Initialize() before it is usable. This, multi-step initialization,  is an anti-pattern IMHO.

I'd rather have a private/protected constructor and a factory method that setups everything and returns the Listener in a unique ptr. 

I.e. exactly what your "CreateFeedbackShooter" is doing.  If you need help to get this done I can sit down with you.

File: src/Listener.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      virtual ~Listener() { Reset(); };
+      Listener(const Listener&) = delete;
+      Listener& operator=(const Listener&) = delete;
+      bool Initialize();
+      void Reset();
+      bool NotificationReceived();
+      bool SendConfirmation();
+      std::unique_ptr<Rifle> CreateFeedbackShooter();
```
Comment (KjellKod) [inline src/Listener.h]: Why is this public. Only the public API should be public. Everything else should be private or protected.

File: src/Notifier.cpp [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+#include <StopWatch.h>
+
+/*
+ * Initialize the mutex-guarded Shotgun-Alien
+ *    queue by "aiming" it at the Notifier
+ *    ipc file
+ *
+ * @return bool, whether or not both queues are initialized
+ */
+bool Notifier::Initialize(const size_t handshakeCount) {
```
Comment (KjellKod) [inline src/Notifier.cpp]: See comments above regarding multiple step initializations.

File: src/Notifier.h [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+#include <mutex>
+#include <memory>
+
+class Shotgun;
+class Vampire;
+
+class Notifier {
+   public:
+      Notifier() = delete;
+      Notifier(const std::string& notifierQueue, const std::string& handshakeQueue)
```
Comment (KjellKod) [inline src/Notifier.h]: make this private

**Comment by KjellKod** [comment]
Overall good. Some cleanup for the API is needed.
0) Please add this to the wiki and point out your description to that in the header of Listener and Notifier
1) Make construction easy, single step for the user
2) Only have public API for what is needed to have public

File: src/Listener.cpp (hunk 1 above)
Comment (KjellKod) [inline src/Listener.cpp]: yes. it's a unique_ptr that gets replaced. Anything old will be deleted.

**Comment by KjellKod** [comment]
:+1:

**Pull request merged**